service/src/main/java/bio/terra/workspace/service/workspace/WorkspaceService.java

```diff
@@ -178,16 +179,33 @@ public List<Workspace> listWorkspaces(
             () -> samService.listWorkspaceIdsAndRoles(userRequest), "listWorkspaceIds");
     return workspaceDao
         .getWorkspacesMatchingList(samWorkspaceIdsAndRoles.keySet(), offset, limit)
-        // For each workspace, set highest role from SAM
         .stream()
+        // For each workspace, set highest role from SAM
         .map(
             workspace -> {
               workspace.setHighestRole(samWorkspaceIdsAndRoles.get(workspace.getWorkspaceId()));
               return workspace;
             })
+        .filter(WorkspaceService::removeDataSourceWorkspacesIfReader)
         .collect(Collectors.toList());
   }
 
+  // WSM stores data sources (such as AMP-PD) as workspaces. (This is an implementation detail;
+  // Terra users are not aware that data sources are workspaces.) An AMP-PD researcher would not
+  // expect to see AMP-PD workspace in their workspace list. So hide data source workspaces, only
+  // for readers. Editors and owners need to see AMP-PD workspace, so they can edit it.
+  @Traced
+  private static boolean removeDataSourceWorkspacesIfReader(Workspace workspace) {
```

Comment:
Access level should be orthogonal to visibility of data sources (which are now called datasets?). This would prevent someone with read access to the dataset from listing datasets they can read. It doesn't make sense to me.

We should be listing data sources if we are asked to list them and not list them by default.
I think we should add a query param to the listWorkspaces to control what workspace types the caller gets in the list.

Reply:
I generally agree with the logic, in that its probably what we want in the short term.

But I think this application-specific magic logic that WSM shouldn't be concerned with.  This seems better suited to a higher layer, that can apply filters as query params or on the result.

This change only works if we add a distinct listDatasets API, when we could instead add params to listWorkspaces.  I feel like the later is more in the spirit of keeping the WSM API agnostic to what a "workspace" contains.

Reply:
Discussed offline, agree this shouldn't be in WSM, even in short term. Closing PR.